File: algorithms/leetcode/020-valid-parentheses/valid-parentheses.js
const isValid = s => {
    const st = [];

    const brackets = {
        '(': ')',
        '[': ']',
        '{': '}'
    };

    for (let i = 0; i < s.length; i++) {
        if (st.length && !brackets[s[i]]) {
            const want = brackets[st.pop()];

            if (want !== s[i]) {
                return false;
            } else {
                continue;
            }
        }

        st.push(s[i]);
    }

    return st.length === 0;
};

File: leetcode/020-valid-parentheses/ValidParentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> opened = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(') {
                opened.push(')');
            } else if (ch == '{') {
                opened.push('}');
            } else if (ch == '[') {
                opened.push(']');
            } else if (opened.isEmpty() || ch != opened.pop()) {
                return false;
            }
        }
        
        return opened.size() == 0;
    }
}
